enhancements/results-forwarding.md

```diff
@@ -0,0 +1,509 @@
+  store. e.g. a relational database or a cloud-managed database.
+
+The Compliance Operator also stores raw results (currently ARF) in a persistent
+volume claim by default. While this is useful, not everyone has access to
+persistent storage. This issue is apparent in shared testing clusters where
+automation attempts to delete the Compliance Operator namespace, only to have
+the request fail because of a failed volume allocation.
+
+#### Gathering Evidence
+
+ARF files contain a lot of information about the state of the system where the
+compliance scan was effectuated. We could take the view that this is "evidence"
+gathered for the system in check. With this in mind, we can generalize and
+expand the concept to allow for forwarding this evidence to an alternative
+object store, e.g. S3.
+
+In the future, we could expand the "evidence" gathered, e.g. for Kubernetes
+compliance scans we could simply compress the gathered objects the
+`api-resource-collector` downloads and send that too.
+
+While evidence is important to consider, we're going to dedicated a separate
+enhancement to that problem. This particular enhancement is focused on
+forwarding results.
+
+### Goals
+
+* Provide a generic way to implement compliance result forwarding for
+  Compliance Operator. This should be flexible enough to accommodate the
+  current implementation (CRDs and PVCs), while allowing for custom forwarding
+  implementations.
+
+* Define and implement a stable API for forwarder implementations.
+
+* Implement an interface for forwarding result CRDs
+
+* Implement an interface for forwarding remediation CRDs
+
+* Implement a switch to disable in-cluster storage if forwarding is enabled and valid
+
+### Non-Goals
+
+* Deprecate the current CRD/PVC approach: There is still a use-case for this.
+
+* Expand the "evidence" that's currently gathered by the Compliance Operator. This is
+  needed but should be done as part of a separate initiative.
+
+* Implement an archiving solution. That's being discussed as a separate
+[enhancement](https://github.com/ComplianceAsCode/compliance-operator/pull/66).
+
+## Proposal
+
+The Compliance Operator will forward compliance results and remediation
+recommendations if the supplied forwarding endpoint is valid. The Compliance
+Operator will also continue to use CRDs and PVCs to store results,
+remediations, and evidence as it does today.
+
+A user may explicitly disable in-cluster storage if forwarding is enabled. The
+Compliance Operator will not support disabling PVC storage without a valid
+forwarding endpoint.
+
+This will require changing the **aggregator** to always forward via a gRPC API
+to a configured implementation. In this case, the **aggregator** will act as a
+gRPC client. The **resultserver** then becomes a gRPC server, subject to the
+evidence forwarding provider and will be renamed **evidence-persistor** and
+responsible for writing results.
```

Comment:
In this case - what does the `aggregator` currently use to pass results to the `resultserver`?

Or are we just saying that the `aggregator` will be a gRPC client. An example server implementation would be `resultserver` or rhmdnd/compserv?

Reply:
The former.

Currently, the aggregator parses all the results coming as ConfigMaps from the nodes.

It "aggregates" them together, forming groups from the different machine pools. e.g. one result is applicable to a set of nodes, not just one node. This is done also to detect inconsistencies in pools: "One node deviated from the benchmark configuration, it should be fixed."

Finally, it takes that aggregated result and creates the CRDs based on this.

The idea is to change the last step from the model and turn the aggregator into a gRPC client that always forwards. This way, that code becomes more general and we can then focus on writing server implementations: Be it the default one that creates CRDs, or compserv that centralizes the results.

Reply:
Thanks for the additional details. I attempted to clarify that in the wording.